### PR TITLE
[Docs] Rewrite 'Trees' section

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Rewrite 'Trees' section as part of the documentation week.

WIP by @Axel-Naumann, @jalopezg-r00t, and @etejedor.

### CODE REVIEW

File: manual/trees/index.md [hunk 1]
```diff
@@ -7,178 +7,187 @@ toc: true
...
-*............................................................................*
-...
-...
-...
+float var;
+tree->Branch("branch0", &var);
+{% endhighlight %}
+
+{% highlight Python %}
+# Provide a one-element array, so ROOT can read data from this memory.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:140]: Maybe clarify a bit this comment.

File: manual/trees/index.md [hunk 2]
```diff
@@ -7,178 +7,187 @@ toc: true
...
+TNamed var;
+const int splitLevel = 99; // "all the way"
+tree->Branch("branch0", &var, splitlevel);
+{% endhighlight %}
+<br/>
+**3. Branches holding `std::vector`, `std::array`, `std::list`, etc**
 
+Both top-level branches (those created by a call to `TTree::Branch()`) and branches created by splitting data members can hold collections such as `std::vector`, `std::array`, `std::list`, or `std::map`.
+Splitting can traverse through collections:
+if a member is a `vector<X>`, the tree can split `X` into sub-branches, too.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:182]:
```suggestion
if a member is a `std::vector<X>`, the tree can split `X` into sub-branches, too.
```

File: manual/trees/index.md [hunk 3]
```diff
@@ -190,13 +199,60 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+   // "branchName" in tree entry `iEntry`
+   printf("%d\n", variable);
+}
+{% endhighlight %}
+
+In Python you can simply use the branch name as an attribute on the tree:
+
+{% highlight Python %}
+for entry in tree:
+   print(entry.branchName)
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:238]: I think iteration over entries is performed this way.  @etejedor can you confirm?

File: manual/trees/index.md [hunk 4]
```diff
@@ -190,13 +199,60 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+{% endhighlight %}
+
+
+### Selecting a subset of branches to be read
+
+You can select or deselect branches from being read by `GetEntry()` by calling [`TTree::SetBranchStatus()`](https://root.cern/doc/master/classTTree.html#aeca53bcd4482de1d883d259df6663920).
+It is vividly recommended to only read the branches actually needed:
+`TTree` is optimized for exactly this use case, and most analyses will only need a fraction of the available branches.
+
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:247]: Maybe add a simple example?

**Review by jalopezg-r00t — COMMENTED** [review]
See comments.

File: manual/trees/index.md [hunk 5]
```diff
@@ -7,178 +7,187 @@ toc: true
...
 
 > **RNTuple**
 >
-> [RNTuple](https://root.cern/doc/master/md_tree_ntuple_v7_doc_README.html){:target="_blank"} (for N-tuple and nested tuple) is the experimental evolution of {% include ref class="TTree" %} columnar data storage. `RNTuple` introduces new interfaces that are more robust.
+> [RNTuple](https://root.cern/doc/master/md_tree_ntuple_v7_doc_README.html){:target="_blank"} is the experimental evolution of {% include ref class="TTree" %} columnar data storage. {% include ref class="RNTuple" namespace="ROOT::Experimental" %} introduces robust interfaces, a high-performance storage layout, and an asynchronous, thread-safe scheduling.
 
-## Tree classes
+> **RDataFrame**
+>
+> TODO: tell people to use RDF, not TTree
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md:30]: We will word that in an even stronger way!

File: manual/trees/index.md (hunk 3 above)
Comment (Axel-Naumann) [reply manual/trees/index.md:238]: He confirms!

File: manual/trees/index.md [hunk 6]
```diff
@@ -7,178 +7,145 @@ toc: true
...
 
-- [TChain](https://root.cern/doc/master/classTChain.html){:target="_blank"}: A list of ROOT files containing `TTree` objects.
+### Branches and leaves
 
+A tree consists of a list of independent columns, called branches. A branch can contain values of any fundamental type, C++ objects known to ROOT's type system, or collections of those.
+Branches are represented by {% include ref class="TBranch" %} and its derived classes.
 
-## Working with trees
+While `TBranch` represent structure, objects inheriting from {% include ref class="TLeaf" %} contain the actual data.
+Originally, any columnar data was contained in a `TLeaf`; these days, some of the `TBranch`-derived classes contain data themselves, such as {% include ref class="TBranchElement" %}.
```
Comment (pcanal) [inline manual/trees/index.md:48]: > `contain data`

Might be an over simplifications.  Literally speaking the only entity that contains the data  are:
* the TBasket (in memory, uncompressed)
* the region of TFile corresponding to this TBasket (on disk, compressed)
* the user variable (in memory.

And it is always the TBranch that holds the information about the baskets.

The distinction between branches and leaves is "whom" hold the address of the memory corresponding to the user variable that hold the data or "whom" can give access to the user variable address and type information (`GetAddress`, `GetValue`).

so maybe `give access to the data` is more accurate.

File: manual/trees/index.md [hunk 7]
```diff
@@ -7,178 +7,145 @@ toc: true
...
+When writing a `TTree` you first want to create a `TFile`
+(see → [ROOT files]({{ '/manual/root_files' | relative_url }}).
+Then construct the `TTree` to be stored in the file; we will later add branches to the tree.
 
 _**Example**_
 
 {% highlight C++ %}
-   TTree t("MyTree","Example Tree");
+std::unique_ptr<TFile> myFile( TFile::Open("file.root", "RECREATE") );
+auto tree = std::make_unique<TTree>("tree", "The Tree Title");
```
Comment (pcanal) [inline manual/trees/index.md:64]: The "unique" part is misleading since there is shared owership between the user and the `TFile`.  A raw pointer is likely adequate here as *usually* the user should not have to worry about deleting the TTree (but they can if they want to).

File: manual/trees/index.md [hunk 8]
```diff
@@ -7,178 +7,145 @@ toc: true
...
-      &staff.flag,&staff.age);
-      sscanf(&line[13],"%d%d%d%d",&staff.service,&staff.children,
-      &staff.grade,&staff.step);
-      sscanf(&line[24],"%d%d%d",&staff.nation,&staff.hrweek,
-      &staff.cost);
-      tree->Fill();
-   }
+> **Note**
+>
+> The objects / variables used to create branches must not be destroyed until the `TTree` is deleted or `TTree::ResetBranchAddress()` is called.
```
Comment (pcanal) [inline manual/trees/index.md:83]:
```suggestion
> The objects *and* variables used to create branches must not be destroyed until the `TTree` is deleted or `TTree::ResetBranchAddress()` is called.
```

File: manual/trees/index.md [hunk 9]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+"Friend trees" are added by calling [TTree::AddFriend()](https://root.cern/doc/master/classTTree.html#a321f2684de145cfcb01cabfce69ea910){:target="_blank"}.
+Adding another tree called `T1` as a friend tree will make the branch `X` of `T1` available as both `T1.X` and - if `X` does not exist in the original tree - as `X`.
+
+Friend trees are expected to have at least as many entries as the original tree.
+The order of the friend tree's entries must preserve the entry order of the original tree.
+
+> **Note**
+>
+> Especially when processing the original tree in parallel to generate the friend tree, the entries might be written out in an undefined order.
+> This can cause the entries of the friend tree to be in a different order than the entries of the original tree.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:293]: I would rephrase this a bit.  Maybe something along the lines of
```suggestion
> Care must be taken to ensure that the order of entries in the primary tree matches friends' entries. This is especially relevant when processing a tree in parallel to generate a friend tree, as the entries might be written out in an undefined order (misaligned entries).
```

File: manual/trees/index.md [hunk 10]
```diff
@@ -7,178 +7,145 @@ toc: true
...
-*Entries :     3354 : Total  Size=      14067 bytes  One basket in memory    *
-*Baskets :        0 : Basket Size=      32000 bytes  Compression=   1.00     *
-*............................................................................*
-...
-...
-...
-{% endhighlight %}
+If the split level is set to 0, there is no splitting: all data members are stored in the same branch.
+Data members can also be configured to be non-split as part of the dictionary; see → [I/O]({{ '/manual/io' | relative_url }}).
+The default split level is 1; a split level of 99 means to split all members at any recursion level.
```
Comment (pcanal) [inline manual/trees/index.md:116]:
```required_change
The default split level is 99 which means to split all members at any recursion level.
```

File: manual/trees/index.md [hunk 11]
```diff
@@ -7,178 +7,145 @@ toc: true
...
 
-### Creating a tree from a folder structure
+_Pointers_
 
-You can build a folder structure and create a tree with branches for each of the sub-folders.
+While references `X &` are not supported as member types, pointers are.
+If the pointer is non-null, ROOT stores the object pointed to (pointee).
+If multiple pointers point to the same object during one `TTree::Fill()` operation, that pointee will only be stored once; upon reading, all pointers will again point to the same object.
```
Comment (pcanal) [inline manual/trees/index.md:122]: Technically that is during on `TBranch::Fill`.  For example if we have (mostly likely indirectly rather than this improbable simplification :) ):
```
struct SomeClass {
   Other *fPointer;
   Ohter *fAnotherPointer; 
};
```
and `fPointer` and `fAnotherPointer` points to the same address **and** the struct `SomeClass` is stored with split level 1 or higher, then the pointee object will be stored twice (and more annoyingly when retrieving the object, `fPointer` and `fAnotherPointer` will not point to two distinct objects .... hence the need/recommendation to use references (like TRef) or indices instead of pointers.

File: manual/trees/index.md [hunk 12]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
 
 When writing a {% include ref class="TTree" %} to a ROOT file and if the ROOT file size reaches the value stored in the [TTree::GetMaxTreeSize()](https://root.cern/doc/master/classTTree.html#aca38baf017a203ddb3119a9ab7283cd9){:target="_blank"}, the current ROOT file is closed and a new ROOT file is created. If the original ROOT file is named `myfile.root`, the subsequent ROOT files are named `myfile_1.root`, `myfile_2.root`, etc.
```
Comment (pcanal) [inline manual/trees/index.md:158]:
```suggestion
When filling a {% include ref class="TTree" %} into a ROOT file and if the ROOT file size reaches the value stored in the [TTree::GetMaxTreeSize()](https://root.cern/doc/master/classTTree.html#aca38baf017a203ddb3119a9ab7283cd9){:target="_blank"}, the current ROOT file is closed and a new ROOT file is created. If the original ROOT file is named `myfile.root`, the subsequent ROOT files are named `myfile_1.root`, `myfile_2.root`, etc.
```

File: manual/trees/index.md [hunk 13]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
 
-**Autosave**
+_Autosave_
```
Comment (pcanal) [inline manual/trees/index.md:160]: We should consider adding text about `AutoFlush` and clusters around here.

File: manual/trees/index.md [hunk 14]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+myFile = ROOT.TFile.Open("file.root")
+myTree = myFile.TreeName
+for entry in myTree:
+   print(entry.branchName)
+{% endhighlight %}
+
+
+### Selecting a subset of branches to be read
+
+You can select or deselect branches from being read by `GetEntry()` by calling [`TTree::SetBranchStatus()`](https://root.cern/doc/master/classTTree.html#aeca53bcd4482de1d883d259df6663920).
```
Comment (pcanal) [inline manual/trees/index.md:208]:
```suggestion
You can select or deselect branches from being read by `TTree::GetEntry()` by calling [`TTree::SetBranchStatus()`](https://root.cern/doc/master/classTTree.html#aeca53bcd4482de1d883d259df6663920).
```

File: manual/trees/index.md [hunk 15]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+{% endhighlight %}
 
-`Autosave` gives you the option to save all branch buffers every n byte. It is recommended to use `Autosave` for large acquisitions. If the acquisition fails to complete, you can recover the ROOT file and all the contents since the last `Autosave`.
 
-- Use the [TTree::SetAutosave()](https://root.cern/doc/master/classTTree.html#a76259576b0094536ad084cde665c13a8){:target="_blank"} method to set the number of bytes between `Autosave`.
+### Selecting a subset of entries to be read
 
-You can also use [TTree::SetAutosave()](https://root.cern/doc/master/classTTree.html#a76259576b0094536ad084cde665c13a8){:target="_blank"} in the acquisition loop every n entry.
+To process only a selection of tree entries, you can use a {% include ref class="TEntryList" %}.
+First you inserting the tree entry numbers you want to process into the `TEntryList`.
```
Comment (pcanal) [inline manual/trees/index.md:235]:
```suggestion
First you insert the tree entry numbers you want to process into the `TEntryList`.
```

File: manual/trees/index.md [hunk 16]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+   # all entries will have missingET < 100
+{% endhighlight %}
+
+## Appending `TTree`s as a `TChain`
+
+In high energy physics you always want as much data as possible.
+But it's not nice to deal with files of multiple terabytes.
+ROOT allows to to split data across multiple files, where you can then access the files' tree parts as one large tree.
+That's done through {% include ref class="TChain" %}, which inherits from {% include ref class="TTree" %}:
+it wants to know the name of the trees in the files (has to be the same throughout), and the file names, and will act as if it was a huge, continuous tree:
```
Comment (pcanal) [inline manual/trees/index.md:262]:
```suggestion
it wants to know the name of the trees in the files, and the file names, and will act as if it was a huge, continuous tree:
```
> (has to be the same throughout),

TChain actually works properly with names change as long as it is specified when adding the file to the chain.

File: manual/trees/index.md [hunk 17]
```diff
@@ -190,13 +157,168 @@ The `TTree::Write()` method is needed to write the ROOT file header.
...
+chain = ROOT.TChain("CommonTreeName")
+if chain.Add("data_*.root") != 12:
+   print("Expected to find 12 files!")
+# Use `chain` as if it was a `TTree`
+{% endhighlight %}
+
+## Widening a `TTree` through friends
+
+Trees are usually written just once.
+While updating an existing tree is non-trivial, extending it with additional branches, potentially an "improved" version of an original branch, is trivial.
```
Comment (pcanal) [inline manual/trees/index.md:283]:
```suggestion
While updating an existing tree is non-trivial, extending it with additional branches, potentially an "improved" version of an original branch, is straighforward.
```

File: manual/trees/index.md (hunk 7 above)
Comment (Axel-Naumann) [reply manual/trees/index.md:64]: I think the key is:

> but they can if they want to

That means we should use the best practices pattern, and that's a stack variable - which is awkward because of ROOT's memory management requirements, and thus the secon best option is a `unique_ptr`. Do you see a *problem* with that, or is it "just" misleading?

File: manual/trees/index.md (hunk 1 above)
Comment (Axel-Naumann) [reply manual/trees/index.md:98]: Discusses with Enric, clarification can quickly go out of control here :-) and we agree this is "good enoough".

File: manual/trees/index.md (hunk 7 above)
Comment (pcanal) [reply manual/trees/index.md:64]: it is just misleading.

File: manual/trees/index.md (hunk 6 above)
Comment (etejedor) [reply manual/trees/index.md:48]: Done!

File: manual/trees/index.md (hunk 10 above)
Comment (etejedor) [reply manual/trees/index.md:116]: Done!

File: manual/trees/index.md (hunk 11 above)
Comment (etejedor) [reply manual/trees/index.md:122]: Thanks for the clarification! Addressed.

File: manual/trees/index.md (hunk 12 above)
Comment (etejedor) [reply manual/trees/index.md:158]: We rephrased this, we think it makes more sense now.

File: manual/trees/index.md [hunk 18]
```diff
@@ -113,13 +132,16 @@ The recusion level of nested splitting is called the "split level"; it can be co
...
 Data members can also be configured to be non-split as part of the dictionary; see → [I/O]({{ '/manual/io' | relative_url }}).
-The default split level is 1; a split level of 99 means to split all members at any recursion level.
+The default split level of 99 means to split all members at any recursion level.
 
 _Pointers_
 
 While references `X &` are not supported as member types, pointers are.
 If the pointer is non-null, ROOT stores the object pointed to (pointee).
-If multiple pointers point to the same object during one `TTree::Fill()` operation, that pointee will only be stored once; upon reading, all pointers will again point to the same object.
+If multiple pointers within the same branch point to the same object during one `TBranch::Fill()` operation (as invoked by `TTree::Fill()`), that pointee will only be stored once; upon reading, all pointers will again point to the same object.
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md:141]: Done

File: manual/trees/index.md [hunk 19]
```diff
@@ -151,18 +173,25 @@ Use [TTree:Fill()](https://root.cern/doc/master/classTTree.html#a00e0c422f5e4f6e
...
-The `TTree::Write()` method is needed to write the ROOT file header.
+If due to the data written during `TTree::Fill()`, the file's size increases beyond [TTree::GetMaxTreeSize()](https://root.cern/doc/master/classTTree.html#aca38baf017a203ddb3119a9ab7283cd9){:target="_blank"}, the current ROOT file is closed and a new ROOT file is created.
+For an original ROOT file named `myfile.root`, the subsequent ROOT files are named `myfile_1.root`, `myfile_2.root`, etc.
 
-When writing a {% include ref class="TTree" %} to a ROOT file and if the ROOT file size reaches the value stored in the [TTree::GetMaxTreeSize()](https://root.cern/doc/master/classTTree.html#aca38baf017a203ddb3119a9ab7283cd9){:target="_blank"}, the current ROOT file is closed and a new ROOT file is created. If the original ROOT file is named `myfile.root`, the subsequent ROOT files are named `myfile_1.root`, `myfile_2.root`, etc.
+_AutoFlush_
 
-_Autosave_
+The tree can flush its data (i.e. its baskets) to file when reaching a given cluster size, thus closing the cluster.
+By default this happens every 30MB.
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md:222]: Applied

File: manual/trees/index.md (hunk 16 above)
Comment (Axel-Naumann) [reply manual/trees/index.md:328]: Clarified

File: manual/trees/index.md (hunk 13 above)
Comment (Axel-Naumann) [reply manual/trees/index.md]: Addressed

File: manual/trees/index.md [hunk 20]
```diff
@@ -7,178 +7,187 @@ toc: true
...
+>
+>  {% include tutorial name="tree3" %}
 
-_**Example: A simple tree**_
+Adding a branch is often not possible because the tree is a read-only file and you do not have permission to save the modified tree with the new branch. Even if you do have the permission, you risk loosing the original tree with an unsuccessful attempt to save the modification. Since trees are usually large, adding a branch could extend it over the 2 GB limit. In this case, the attempt to write the tree fails, and the original data is may also be corrupted. In addition, adding a branch to a tree enlarges the tree and increases the amount of memory needed to read an entry, and therefore decreases the performance.
 
-The following script builds a {% include ref class="TTree" %} from an ASCII file containing
-statistics about the staff at CERN. Both, `staff.C` and `staff.dat` are in available in
-`$ROOTSYS/tutorials/tree`.
+For these reasons ROOT offers the concept of friends for trees (and chains) by adding a branch manually with [TTree::AddFriend()](https://root.cern/doc/master/classTTree.html#a011d362261b694ee7dd780bad21f030b){:target="_blank"}.
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md]: Addressed (differently)

File: manual/trees/index.md [hunk 21]
```diff
@@ -7,178 +7,187 @@ toc: true
...
-{% endhighlight %}
+### Adding friends to trees
 
-It creates a tree with the title `Example Tree`.
+> Tutorial
+>
+>  {% include tutorial name="tree3" %}
 
-_**Example: A simple tree**_
+Adding a branch is often not possible because the tree is a read-only file and you do not have permission to save the modified tree with the new branch. Even if you do have the permission, you risk loosing the original tree with an unsuccessful attempt to save the modification. Since trees are usually large, adding a branch could extend it over the 2 GB limit. In this case, the attempt to write the tree fails, and the original data is may also be corrupted. In addition, adding a branch to a tree enlarges the tree and increases the amount of memory needed to read an entry, and therefore decreases the performance.
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed

File: manual/trees/index.md [hunk 22]
```diff
@@ -7,178 +7,187 @@ toc: true
...
-- Use the {% include ref class="TTree" %} constructor to create a tree.
+### Widening a `TTree`s through friends
 
-_**Example**_
+[TODO](https://TRIGGER CI!)
 
-{% highlight C++ %}
-   TTree t("MyTree","Example Tree");
-{% endhighlight %}
+### Adding friends to trees
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed

File: manual/trees/index.md [hunk 23]
```diff
@@ -44,8 +44,27 @@ when reading back a `TTree` entry, it will write the values it read from storage
...
+Once the buffer is full, it gets compressed; the compressed buffer is called _basket_.
+These baskets are written into the ROOT file.
+Branches with more data per tree entry will fill more baskets than branches with less data per tree entry.
+Conversely, baskets can hold many tree entries if their branch stores only a few bytes per tree entry.
+This means that generally, all baskets - also of different branches - will contain data of different tree entry ranges.
+
+To allow more efficient pre-fetching and better chunking of tree data stored in ROOT files, TTree groups baskets into clusters, for a defined range of tree entry indices.
+Trees will close baskets that are not yet full when reaching the tree entry at a cluster boundary.
+
+TTree finds the baskets for a given entry for a given branch by means of a header stored to file.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:62]:
```suggestion
TTree finds the baskets for a given entry for a given branch by means of a header stored in the file.
```

File: manual/trees/index.md [hunk 24]
```diff
@@ -113,13 +132,16 @@ The recusion level of nested splitting is called the "split level"; it can be co
...
+The default split level of 99 means to split all members at any recursion level.
 
 _Pointers_
 
 While references `X &` are not supported as member types, pointers are.
 If the pointer is non-null, ROOT stores the object pointed to (pointee).
-If multiple pointers point to the same object during one `TTree::Fill()` operation, that pointee will only be stored once; upon reading, all pointers will again point to the same object.
+If multiple pointers within the same branch point to the same object during one `TBranch::Fill()` operation (as invoked by `TTree::Fill()`), that pointee will only be stored once; upon reading, all pointers will again point to the same object.
+
+For the general case, indices into object collections could be persistified instead of pointers.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:143]:
```suggestion
For the general case, indices into object collections could be persisted instead of pointers.
```

File: manual/trees/index.md [hunk 25]
```diff
@@ -147,22 +169,66 @@ If the split level is `TTree::kSplitCollectionOfPointers` then the pointees will
...
+# Now write the header
+tree.Write()
+{% endhighlight %}
 
-- Use the [TTree::Write()](https://root.cern/doc/master/classTTree.html#af6f2d9ae4048ad85fcae5d2afa05100f){:target="_blank"} method to write the tree into a ROOT file.
+_AutoFlush_
 
-The `TTree::Write()` method is needed to write the ROOT file header.
+The tree can flush its data (i.e. its baskets) to file when reaching a given cluster size, thus closing the cluster.
+By default this happens approximatively every 30MB of compressed data.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:222]:
```suggestion
By default this happens approximately every 30MB of compressed data.
```

File: manual/trees/index.md [hunk 26]
```diff
@@ -320,19 +385,16 @@ void treeWithFriend() {
 
 ## Examining a tree
 
-### Printing the summary of a tree
+Different ways to examine the tree structure and content exist, from text to graphics.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:388]:
```suggestion
ROOT offers different ways to examine tree structure and its contents, from text to graphics.
```

File: manual/trees/index.md [hunk 27]
```diff
@@ -517,6 +511,44 @@ The scatterplot is drawn.
 Note that not each `(x,y) point on a scatterplot represents two values in your N−tuple. In fact, the scatterplot is a grid and each square in
 the grid is randomly populated with a density of dots that’s proportional to the number of values in that grid.
 
+### Indexing trees
+
+- Use [TTree::BuildIndex()](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bd0b06eea6c12b998){:target="_blank"} method to build an index table using expressions depending on the value in the leaves.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:516]:
```suggestion
- Use the [TTree::BuildIndex()](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bd0b06eea6c12b998){:target="_blank"} method to build an index table using expressions depending on the value in the leaves.
```

File: manual/trees/index.md [hunk 28]
```diff
@@ -517,6 +511,44 @@ The scatterplot is drawn.
...
+
+The index is built in the following way:
+- A pass on all entries is made like in [TTree::Draw()](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bTTree::Draw()d0b06eea6c12b998){:target="_blank"}.
+- `var1` = `majorname`
+- `var2` = `minorname`
+- `sel = 231` × _majorname_ + _minorname_
+- For each entry in the tree the `sel` expression is evaluated and the result array is sorted into `fIndexValues`.
+
+Once the index is calculated, an entry can be retrieved with
+[TTree::GetEntryWithIndex(majornumber, minornumber)](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bTTree::Draw()d0b06eea6c12b998){:target="_blank"}.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:526]: Broken link?

File: manual/trees/index.md [hunk 29]
```diff
@@ -517,6 +511,44 @@ The scatterplot is drawn.
...
+   tree.BuildIndex("Run","Event");
+
+// To read entry corresponding to Run=1234 and Event=56789.
+   tree.GetEntryWithIndex(1234,56789);
+   {% endhighlight %}
+
+Note that `majorname` and `minorname` can be expressions using original tree variables e.g., `"run-90000"` or `"event +3*xx"`.
+
+In case an expression is specified, the equivalent expression must be computed when calling
+[TTree::GetEntryWithIndex(majornumber, minornumber)](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bTTree::Draw()d0b06eea6c12b998){:target="_blank"}.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:541]: Broken link?

File: manual/trees/index.md [hunk 30]
```diff
@@ -517,6 +511,44 @@ The scatterplot is drawn.
...
+To build an index with only `majorname`, specify `minorname="0"` (default).
+
+Once the index is built, it can be saved with the `TTree` object with `tree.Write()`.
+
+The most convenient place to create the index is at the end of the filling process just before saving the tree header. If a previous index was calculated, it will be redefined by this new call.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:546]:
```suggestion
The most convenient place to create the index is at the end of the filling process just before saving the tree header. If a previous index was available, it will be overridden by this new call.
```

**Review by jalopezg-r00t — COMMENTED** [review]
See comments.

File: manual/trees/index.md [hunk 31]
```diff
@@ -511,42 +511,38 @@ The scatterplot is drawn.
...
+### Indexing a tree
+
+- Use [TTree::BuildIndex()](https://root.cern/doc/master/classTTree.html#a3f6b5bb591ff7a5bd0b06eea6c12b998){:target="_blank"} to build an index table over expressions that depend on the value in the leaves.
+This index is similar to database indexes:
+it allows to quickly determine the tree entry number corresponding to the value of an expression.
+These expressions should be both equality comparable (that is, not use floating point numbers where precision might cause the index lookup to fail) and unique, to make sure you get the tree entry you expect.
+For high-energy physics, a common example could be a combination of run number and event number:
+while each one of them might have duplications, their combination is guaranteed to be unique. 
+
+To build an index, define a major and optionally a minor expression, for instance above `Run` and `Event`.
```
Comment (etejedor) [inline manual/trees/index.md:523]: Says above but the example is below?

File: manual/trees/index.md (hunk 28 above)
Comment (etejedor) [reply manual/trees/index.md:530]: https://root.cern.ch/doc/master/classTTree.html#a8c77a02a1323019375d273299a2c14b1

Comment (etejedor) [reply manual/trees/index.md:530]: Also, I believe we never include the parameters when we link to a method, so I would remove `majornumber, minornumber`

File: manual/trees/index.md (hunk 31 above)
Comment (etejedor) [reply manual/trees/index.md:523]: Should we say they are leaves?

File: manual/trees/index.md (hunk 24 above)
Comment (etejedor) [reply manual/trees/index.md:143]: Persistify is actually the transitive verb here, so we think it's ok.

File: manual/trees/index.md [hunk 32]
```diff
@@ -59,10 +59,11 @@ Branches with more data per tree entry will fill more baskets than branches with
 Conversely, baskets can hold many tree entries if their branch stores only a few bytes per tree entry.
 This means that generally, all baskets - also of different branches - will contain data of different tree entry ranges.
 
-To allow more efficient pre-fetching and better chunking of tree data stored in ROOT files, TTree groups baskets into clusters, for a defined range of tree entry indices.
+To allow more efficient pre-fetching and better chunking of tree data stored in ROOT files, TTree groups baskets into _clusters_.
+A cluster contains all the data of a given event range.
```
Comment (etejedor) [inline manual/trees/index.md:63]: Is it ok to talk about `event range` when so far we've talked about `entries` ?

File: manual/trees/index.md (hunk 31 above)
Comment (pcanal) [reply manual/trees/index.md]: No, they are really expressions (eg. `run%1000`, `event_part_1 * 1000 + event_part_2`)

File: manual/trees/index.md [hunk 33]
```diff
@@ -68,7 +72,17 @@ Only the last one (also accessible as `treename`) knows about all written basket
 
 ### `TNtuple`, the high-performance spread-sheet
 
-For convenience, ROOT also provides the {% include ref class="TNtuple" %} class which is a {% include ref class="TTree" %} that is limited to contain floating-point numbers only.
+For convenience, ROOT also provides the {% include ref class="TNtuple" %} class which is a tree whose branches contain only numbers of type `float`, one per tree entry.
+It derives from {% include ref class="TTree" %} and is constructued with a `;` separated list of column names.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:76]:
```suggestion
It derives from {% include ref class="TTree" %} and is constructed with a list of column names separated by `:`.
```

File: manual/trees/index.md (hunk 32 above)
Comment (Axel-Naumann) [reply manual/trees/index.md]: Fixed thanks!

File: manual/trees/index.md [hunk 34]
```diff
@@ -7,985 +7,543 @@ toc: true
...
+   for (int iEntry = 0; tree->LoadTree(iEntry) >= 0; ++iEntry) {
+      // Load the data for the given tree entry
+      tree->GetEntry(iEntry);
 
+      printf("%d %d\n", variable, variableFriend);
+   }
+{% endhighlight %}
 
-### Using TTree::MakeSelector()
+# We still need to work on the parts below. No point in reviewing yet.
```
Comment (jalopezg-r00t) [inline manual/trees/index.md:400]: If this is going to be merged in its current state, please remove this!
```suggestion
```

**Review by jalopezg-r00t — APPROVED** [review]
As discussed on Friday, the documentation week ended and this should be merged.
It is in a much better state now, but see suggestion before merging.

File: manual/trees/index.md (hunk 34 above)
Comment (etejedor) [reply manual/trees/index.md:400]: Nice catch thanks!

**Comment by etejedor** [comment]
Thank you @jalopezg-r00t for the catch! All comments have been resolved so I'll merge if checks are fine.

**Comment by etejedor** [comment]
Still can't merge, for two reasons:
- There is a link from the histograms section to an example that we removed in trees. To be discussed today in the meeting how to solve that (probably just removing the reference to the example and replacing it with a reference to the trees section).
- We have links to I/O from trees but I/O still doesn't exist. To be discussed how to proceed (I'd perhaps comment out those links for now to make it possible to merge trees).

**Review by pcanal — APPROVED** [review]

File: manual/trees/index.md [hunk 35]
```diff
@@ -7,985 +7,542 @@ toc: true
...
-   caption="The variable `Cost` and `Age` with a selection drawn in a histogram."
-%}
+Friend trees are expected to have at least as many entries as the original tree.
+The order of the friend tree's entries must preserve the entry order of the original tree.
 
-Next, the fourth pad is activated and the histogram is drawn with the draw option `surf2`.
-Refer to the {% include ref class="THistPainter" %} class for possible draw options.
+> **Note**
+>
+> Care must be taken to ensure that the order of entries in the primary tree matches friends' entries. This is especially relevant when processing a tree in parallel to generate a friend tree, as the entries might be written out in an undefined order (misaligned entries).
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply manual/trees/index.md:374]: Thanks, addressed.
